### ISSUE TITLE
Feature Request: Identify the publisher of each message

### ISSUE THREAD

**Issue by Britvich** [nats-io/nats-server#1173]
##  Feature Request

Please consider exposing the `client_id` in `MSG`s to uniquely identify the publisher of each message.

#### Use Case:

In multiplayer environments (e.g. MMOs), one of the core problems is cheating. There are many things that can be done to prevent cheating, but the most important one is simple: *don't trust the players*. Always assume the worst – that players *will* try to cheat.

To reduce cheating, each player is required to send their requests to services, which validate and authenticate them, and inform the other interested players what just happened. This is usually called an *authoritative service*, because it has the final say as to what changes occur and who can make them.

With NATS however, when a service receives a `MSG`, there is nothing in the protocol that identifies the publisher of the message, making it impossible to provide an authoritative service.

Therefore, I propose optionally including the `client_id` in the `MSG` protocol, where the `client_id` is unique cross-supercluster.

Here's an example of how one might use this feature to identify the publisher of each message:

1. An *Identity Service* (*IS*) connects to NATS, requesting `client_id`s to be included in `MSG`s
1. An *Authoritative Service* (*AS*) connects to NATS, requesting `client_id`s to be included in `MSG`s
1. A *Player's Client* (*PC*) connects to NATS, and is assigned a unique cross-supercluster `client_id`
1. The *PC* makes a request to the *IS* with their login credentials
1. The *IS* receives the login request `MSG` with the player's `client_id`, and associates their `client_id` with their identity (e.g. user id) in a local dictionary
1. The *PC* makes a request to the *AS*
1. The *AS* receives the request `MSG` with the player's `client_id` and checks in a local dictionary if it has ever seen this `client_id` before, and if not, makes a request to the *IS* for the identity information associated with the player's `client_id`, and adds it to the dictionary

##### NATS SERVERS GENERATE UNIQUE CLIENT IDS

Whenever a server accepts a connection from a client, it generates a unique cross-supercluster `client_id`. Each `client_id` consists of 3 components: a timestamp, sequence number, and an instance (or server) number...

![](https://i2.wp.com/cutecoder.org/wp-content/uploads/2018/03/snowflake-identifier@2x.png)

Adjusting the number of bits devoted to each component of the 64-bit `client_id` determines the lifetime of the id, the number of ids that can be generated per millisecond per server, and the number of servers that can simultaneously generate ids. Here's a suggested setup:

| Component | Bits | Maximum |
| --------- | ---- | ---- |
| Timestamp | 40 | ~34.8 years |
| Sequence | 12 | 4,096 ids/millisecond/server |
| Server | 12 | 4,096 servers |

With this setup, a 4,096-server supercluster could run for ~34.8 years, with each server generating 4,096 unique ids per millisecond (or a total of ~16.8 billion unique ids per second).

##### NATS SERVERS PICK UNIQUE SERVER IDS

Whenever a server starts, it gossips with the existing servers to find an available server id for itself.

##### CONNECT CLIENT_ID OPTION

Connections that want to receive `client_id`s on all `MSG`s indicate this via a new `CONNECT` option:

```json
CONNECT {"client_id":true}
```

##### MSGS INCLUDE CLIENT_ID

If a connection has requested `client_id`s on all `MSG`s, then they are sent as the last number before the payload:

```json
MSG FOO.BAR 9 11 7061295291167849000\r\nHello World\r\n
```

#### Who Benefits From The Change(s)?

Any service that needs to know who published a message benefits. This includes authoritative services that need identity information to validate and authorize requests.

#### Alternative Approaches

Services could require clients to put their `client_id` in the payload of their requests, but they can (and will) just maliciously spoof their identity.

Perhaps NATS servers could also act as the identity service, eliminating the need for writing an external identity service. A new `ID` message:

```json
ID 7061295291167849000\r\n
```

could be used to obtain the identity of a given `client_id`.

**Comment by derekcollison** [comment]
NATS is payload agnostic. We have a way to generate globally unique identifiers, called NUID. You are free to place those in the payload yourself. If you want to make sure they are not tampered with you could use our NKEYS and sign the payload.

https://github.com/nats-io/nuid
https://docs.nats.io/nats-server/configuration/securing_nats/auth_intro/nkey_auth

**Comment by Britvich** [comment]
Wow, closed this request without any discussion?? 

1. I'm not suggesting that the `client_id` be in the *payload*, in fact just the opposite. I'm suggesting it is put in the `MSG` protocol (the stuff before the payload), keeping NATS payload agnostic.
1. Since NATS servers already have 64-bit `client_id`s, I'm suggesting the *existing* `client_id`s be made globally unique, keeping things as simple as possible.
1. I don't understand how putting identity information in the payload and signing it with an NKEY would work...
   1. Since NATS is payload agnostic (which is a good thing), how does a service guarantee that a client put in *their* identity into a `PUB` payload when it receives a `MSG` forwarded from NATS?
   1. If NATS can guarantee the contents of the payload are proper using NKEYS (which I don't see how), can NATS administer millions of identities (NKEYS)?... Adding, changing, and removing them via some protocol?
   1. If a device generates an NKEY (to establish an identity) the first time it connects with NATS, how would a player switch to another device and use that same identity on another device?
 
I'm new to NATS, so maybe it can do all those things, but if not perhaps you may want to reconsider this feature request that solves a very real problem.

**Comment by derekcollison** [comment]
Apologies for being terse.

The client protocol is one of the absolute largest and most disruptive things we can do. We have a 10yr old ecosystem with over 36 different client languages etc. We have plans to upgrade the client protocol at some point, but it is pretty much the same as it was since I designed it 10yrs ago.

Publishers and Subscribers should be decoupled by nature. And we encourage that. We do allow exports and imports for services and streams that can be authorized, meaning you know whoever sent it to you was authorized to do so.

You can also use simple auth via a configuration file as well.

Lots of opportunities to solve the problem with the client protocol as it is today.

**Comment by derekcollison** [comment]
Also note that although we care deeply that a publisher is allowed to send a message on a given subject, we also care deeply that the information about the publisher is obscured and does not leak data. This is for a global utility like NGS where a publisher needs to be authorized but can not be tracked by any metadata.

You can always opt-in to identify the publisher.

**Comment by Britvich** [comment]
No problem on the terseness. I know you are very busy with JetStream, et al... 

I understand the reluctance to make any changes to the client protocol, especially if it breaks existing clients! That's why I suggested clients opt-in to receiving publisher identity using a new `CONNECT` option:

```json
CONNECT {"client_id":true}
```

...so as not to break any existing clients. But yes, as you stated, it would require each of the 36 client implementations to add support for `client_id`s in `MSG`s eventually (very much like `noecho` support was added).

Just so I understand, are you saying you never want NATS to expose *any* information about a publisher (such as some client or session numeric identity), even in a supercluster where all the services are authorized to communicate on a *private* network?

My situation is: I have millions of mobile clients that need to "login" to a supercluster, with each service able to identify who is making each request, without the clients being able to spoof their identity.  I still don't see how the solutions you've listed above do that.

You mentioned "exports and imports can be authorized, meaning whoever sends a message was authorized", which is useful, but not the problem I'm trying to solve. I need to know *who* sent the message, without the sender being able to spoof their identity.

You also said "you can use simple auth via a configuration file", but again services receiving forwarded `MSG`s from NATS have no idea who `PUB`ed the message, unless the publishers sticks their identity in the payload, allowing them to spoof their identity. Also, a "configuration file" seems like a tough way to manage millions of identities that are being added, changed, and removed constantly.

When you said "you can always opt-in to identify the publisher", how would that work?

"Lots of opportunities to solve the problem with the client protocol as it is today." I still don't see how without exposing the publisher identity as this feature request discusses.

**Comment by derekcollison** [comment]
They way we tackled this today was as follows.

We have a service in NGS that reports usage. You send a message (request) to the service and receive a JSON response. Show below.

```
~/Development/go/src/github.com/nats-io/nats.go/examples> nats-req ngs.usage 1h
Published [ngs.usage] : '1h'
Received  [_INBOX.f4ACbaS16RJsjABfKbPvxy.ql3xh5w8] : '{
  "sent": {
    "msgs": 2,
    "bytes": 4
  },
  "received": {
    "msgs": 2,
    "bytes": 234
  },
  "sessions": 1
}'
```

Now what is interesting is that every user has the same subject they send to, so how does the backend service know who is who? The key is that the exported service is on `ngs.usage.*` and authorization is required from the account owner. When the usage services authorizes access it only authorizes a literal subject, like `ngs.usage.<SOMEID>`. In our case we use the public NKEY for the importing account.

Now, what is cool about imports is that you can map them where you want. So when a user signs up for NGS we map everyones import to simply `ngs.usage`. The backend service though receives on the `ngs.usage.<SOMEID>` and its guaranteed to be secure and can not be tampered with. So the backend uses that last token in the subject to lookup the users account usage and return it.

Here is the demos account that responds to usage requests.

```
~/Development/go/src/github.com/nats-io/nats.go/examples> nsc describe account -n demos
╭──────────────────────────────────────────────────────────────────────────────────────╮
│                                   Account Details                                    │
├───────────────────────────┬──────────────────────────────────────────────────────────┤
│ Name                      │ demos                                                    │
│ Account ID                │ ADEMOSSWWUE6EGOSSWDXYVQECHUPOPF4CJOCWCPCTRB6W6L2FEE5Y7IF │
│ Issuer ID                 │ ODSKAYU5JRMXLFXFJU7Y73PNN22ZTXCVTVA26VVLVH2CNM6QV2VBJMIO │
│ Issued                    │ 2019-09-19 20:43:11 UTC                                  │
│ Expires                   │                                                          │
├───────────────────────────┼──────────────────────────────────────────────────────────┤
│ Max Connections           │ 100                                                      │
│ Max Leaf Node Connections │ Not Allowed                                              │
│ Max Data                  │ Unlimited                                                │
│ Max Exports               │ Unlimited                                                │
│ Max Imports               │ Unlimited                                                │
│ Max Msg Payload           │ Unlimited                                                │
│ Max Subscriptions         │ Unlimited                                                │
│ Exports Allows Wildcards  │ True                                                     │
├───────────────────────────┼──────────────────────────────────────────────────────────┤
│ Imports                   │ None                                                     │
╰───────────────────────────┴──────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────╮
│                                       Exports                                        │
├────────────┬─────────┬─────────────┬────────┬─────────────┬──────────────────────────┤
│ Name       │ Type    │ Subject     │ Public │ Revocations │ Tracking                 │
├────────────┼─────────┼─────────────┼────────┼─────────────┼──────────────────────────┤
│ echo       │ Service │ ngs.echo    │ Yes    │ 0           │ -                        │
│ ngs.active │ Stream  │ ngs.active  │ Yes    │ 0           │ N/A                      │
│ usage      │ Service │ ngs.usage.* │ No     │ 0           │ ngs.monitor.usage (100%) │
╰────────────┴─────────┴─────────────┴────────┴─────────────┴──────────────────────────╯
```

And here is my account as an NGS user.

```~/Development/go/src/github.com/nats-io/nats.go/examples> nsc describe account -n derek
╭──────────────────────────────────────────────────────────────────────────────────────╮
│                                   Account Details                                    │
├───────────────────────────┬──────────────────────────────────────────────────────────┤
│ Name                      │ derek                                                    │
│ Account ID                │ ADEREKYKHPGHKP3UOOWSTAVFTYSFSXZELVLTRXOWQ74RG7CI2BFUKOQ3 │
│ Issuer ID                 │ ODSKAYU5JRMXLFXFJU7Y73PNN22ZTXCVTVA26VVLVH2CNM6QV2VBJMIO │
│ Issued                    │ 2019-05-01 17:54:48 UTC                                  │
│ Expires                   │                                                          │
├───────────────────────────┼──────────────────────────────────────────────────────────┤
│ Max Connections           │ Unlimited                                                │
│ Max Leaf Node Connections │ 4                                                        │
│ Max Data                  │ Unlimited                                                │
│ Max Exports               │ Unlimited                                                │
│ Max Imports               │ Unlimited                                                │
│ Max Msg Payload           │ Unlimited                                                │
│ Max Subscriptions         │ Unlimited                                                │
│ Exports Allows Wildcards  │ True                                                     │
├───────────────────────────┼──────────────────────────────────────────────────────────┤
│ Exports                   │ None                                                     │
╰───────────────────────────┴──────────────────────────────────────────────────────────╯

╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                  Imports                                                                   │
├────────────┬─────────┬────────────────────────────────────────────────────────────────────┬──────────────┬─────────┬──────────────┬────────┤
│ Name       │ Type    │ Remote                                                             │ Local/Prefix │ Expires │ From Account │ Public │
├────────────┼─────────┼────────────────────────────────────────────────────────────────────┼──────────────┼─────────┼──────────────┼────────┤
│ ngs.active │ Stream  │ ngs.active                                                         │              │         │ demos        │ Yes    │
│ echo       │ Service │ ngs.echo                                                           │ ngs.echo     │         │ demos        │ Yes    │
│ usage      │ Service │ ngs.usage.ADEREKYKHPGHKP3UOOWSTAVFTYSFSXZELVLTRXOWQ74RG7CI2BFUKOQ3 │ ngs.usage    │         │ demos        │ No     │
╰────────────┴─────────┴────────────────────────────────────────────────────────────────────┴──────────────┴─────────┴──────────────┴────────╯

Hopefully that helps a bit.

Also note that adding things to the CONNECT or INFO block is easy, but what you suggested would be a new protocol directive to deliver a message with a publisher_id attached, that one is the hard one with an ecosystem as large as we have.

**Comment by BrendonW** [comment]
Perhaps I'm misunderstanding the request, but it seems like relying on the messaging layer to provide your player identity authenticity is rife with problems. Since I know the NATS protocol, I'd just fake the client id at the lower level to pretend to be whomever I desire. 

It would seem that the correct solution would be to handle this need in the application layer using digital signatures. Depending on the exact need, JWT may offer a real solution. By having a unique key pair for every client, and signed messages, you can guarantee that messages are unmodified and coming from a particular client, or at worst a clone of the client which is always a possibility if the client is untrustworthy. Making use of message sequence numbers, server provided nonces, etc. along with auto-banning clients that behave strangely can prevent multiple cloned copies from functioning.

TL;DR: The request won't actually solve the authentication problem. Doing this properly is more expensive, involving cryptographic operations, and is best handled at the layer that has the requirements.

**Comment by ripienaar** [comment]
@BrendonW it becomes a lot more useful with accounts or authentication - you cant "fake the protocol" because you need a valid username/password, the intention isnt to get the system to pass what you give unverified, once password is verfied or token authorized or whatever - that information thats been verified is presented in the request.

The thing is, with accounts ALREADY we have unique JWTs and key pairs and all that stuff, the request is to surface that already present details, NATS already ensures messages comes from a particular client unmodified etc. Why not make this information possible? Why force a developer to reinvent that authentication wheel when a lot of good work is already done at the nats layer.

For example - when you use HTTPS and client certificates the webserver makes the authenticated client cert ID available to your code and your code knows it can be trusted because you only get this information once the webserver has done all its checks and balances. That's all that's being asked for her an analogue of that.

That said, Dereks solution seems fine to me as an alternative.

**Comment by derekcollison** [comment]
And to add to @ripienaar's comments, we will start surfacing more things around identity and encryption. Stay tuned. 

But note this is an opt in such that the publisher is the one to put the data into a message versus the NATS system leaking that information at will to any consumer, which is something I do not want to do by default.

**Comment by derekcollison** [comment]
I am going to close for now, but feel free to re-open if needed.

**Comment by cawoodm** [comment]
I also find it weird that NATS doesn’t provide the sender‘s id nor the topic. If you’re subscribing to say all topics (as a logger) you don’t know which channel they were sent on. Adding this to the message payload is a hack: like putting the sender IP address in a TCP packet.

**Comment by ripienaar** [comment]
> I also find it weird that NATS doesn’t provide the sender‘s id nor the topic. If you’re subscribing to say all topics (as a logger) you don’t know which channel they were sent on. Adding this to the message payload is a hack: like putting the sender IP address in a TCP packet.

The message objects do include the subject the message arrived on. That’s key information.

**Comment by derekcollison** [comment]
NATS is meant to be a decoupled publish/subscribe system where producers and consumers are decoupled and separate. That being said we understand at an application level a desire for this information. With headers now we have a new surface area to supply this information without changing the payload. With service latency tracking for exported services we have already been playing with requestors opt-in to share details with the service responder. So I think you will see this type of thing spread out over more parts of NATS.

**Comment by cawoodm** [comment]
The nodeJS nats.subscribe function‘s callback provides only one parameter: the message, no topic. JavaScript lets you add this by double wrapping the callback.

The nats server knows which client sent the message and is IMO laming it’s own adoption in areas where this information is required. Yes, you can pack it into the payload but what if a client misbehaves or is breached? You’ll never know who is flooding the channel.

**Comment by ripienaar** [comment]
I am not a JS person but seems it’s all available?

```
// Subscription/Request callbacks are given multiple arguments:
// - msg is the payload for the message
// - reply is an optional reply subject set by the sender (could be undefined)
// - subject is the subject the message was sent (which may be more specific
//   than the subscription subject - see "Wildcard Subscriptions".
// - finally the subscription id is the local id for the subscription
//   this is the same value returned by the subscribe call.
nc.subscribe('foo', (msg, reply, subject, sid) => {
  if (reply) {
    nc.publish(reply, 'got ' + msg + ' on ' + subject + ' in subscription id ' + sid)
    return
  }
  console.log('Received a message: ' + msg + " it wasn't a request.")
})
```

Conceptually I would say middle ware based systems is deliberately decoupled and message orientated. Knowing the sender suggests a coupling that generally isn’t done

In my software I also need to know the sender - management framework - but it turns out it makes best sense to encode that at my later rather than the middleware so I own all the certificates and AAA etc

**Comment by cawoodm** [comment]
You are correct - I had tried using `callback(msg, foo)` and since `foo` was always undefined I had assumed no other parameters were passed to the callback. RTM. `callback(msg, reply, subject, sid)`!

I am building a distributed computer. Each agent/node/host can post and process tasks. Naturally, when a job is processed, you want to pass it back to the agent which requested it. The agents are decoupled because they don't care where in space and time their tasks are run but at some level you have to couple otherwise data never finds it's way home.

I see two workarounds: agent-specific topics `send.to.agen.<uniqueAgentId>` or adding a layer into the payload.

**Comment by ripienaar** [comment]
Typically reply to topics is for that, we're also busy adding headers support - not sure what the status is in the JS clients, but in server 2.2.0 you wil be able to add headers which lets you get a lot nearer.

By the sounds of it what you really want isnt the originating client name but rather the role of the client that generated the messages, so that you can put it back into that roles work queue rather than back to the individual client. 

sounding more and more to me like perhaps something like your job spec should include a next step key or similar

**Comment by playerx** [comment]
@ripienaar I've the same requirement, I need to find out publisher's user id. (we use nkeys). Now headers are already available and the only thing we are missing is one new flag for eats-server to add publisher's user id (public nkey) to the headers. It's really crucial for our use case and opens lot's of new opportunities in gaming space.

adding user id at the end `send.to.agen.<userId>` as @cawoodm mentioned is a hacky workaround and it will be better to have a normal solution in nats for that, just one flag which can be disabled by default.
